Azure Developer CLI Post (#61)
Azure/Cloud-Native: website/blog/2022-09-29/index.md

    href="https://azure.github.io/Cloud-Native/blog/08-functions-azure" />
</head>

---

Welcome to `Day 29` of #30DaysOfServerless!

---

## What We'll Cover
 * **A new tool has entered the chat:** Azure Developer CLI (`azd`)
 * **(Template) Anatomy 101**: What makes an `azd` template?
 * **End-to-end support:** Move code to cloud in a single step!
 * **Best practices:** Monitor your application and run CI/CD on every commit
 * **Exercise:** Try this yourself or create your own `azd` template
 * **Resources:** For self-study!

![](./img/banner.png)

---

## **A new tool has entered the chat:** What is the Azure Developer CLI (`azd`)?
The Azure Developer CLI (`azd`) is an new, open source tool that makes it quick and simple for you to move your application from your local development environment to Azure while considering your end-to-end developer workflow. You might be familiar with other CLIs that focus on Infrastructure as Code or scaffolding your application but the Azure Developer CLI does all that and more!

The Azure Developer CLI commands are simple, high-level and map to core stages in your developer workflow. Think project initialization/creation, build, deploy, repeat!

> 

Author: Insert graphic here.

### Application Monitoring
When we provisioned all of our infrastructure, we also set up application monitoring via a Bicep file in our `.infra/` directory that spec'd out an Application Insights dashboard. By running `azd monitor` we can see the dashboard with live metrics that was configured for the application. 
![](img/metrics.png)

We can also navigate to the Application Dashboard by clicking on the resource group name, where you can set a specific refresh rate for the dashboard, and see usage, reliability, and performance metrics over time.
![](img/dashboard.png)

I don't know about everyone else but I have spent a ton of time building out similar dashboards. It can be super time-consuming to write all the queries and create the visualizations so this feels like a real time saver.

### CI/CD
Finally let's talk about setting up CI/CD! This might be my favorite `azd` feature. As I mentioned before, the Azure Developer CLI has a command, `azd pipeline config`, which uses the files in the `.github/` directory to set up a GitHub Action. More than that, if there is no upstream repo, the Developer CLI will actually help you create one.

> **@savannah to finish**

Author: Add screenshots after 9/14

![](img/arch-diagram.png)

Let's talk about the files in terms of their purpose:

### Azure Developer CLI-specific files
  - `azure.yaml` - contains metadata that describes the application (Azure hosts, languages, template name) and serves as an entrypoint for functionality in Visual Studio Code (oh yeah, you can also use `azd` using [a VS Code extension](https://aka.ms/azure-dev/vscode-ext?source=serverless-september)!).

### Application development support (code, run, debug, test)
  - `.devcontainer/` - support for if you're [writing code in a container](https://code.visualstudio.com/docs/remote/containers)
  - `.vscode/` - support for local development in Visual Studio Code via [launch.json](https://code.visualstudio.com/docs/editor/debugging) (for debugging) and [tasks.json](https://code.visualstudio.com/docs/editor/tasks) (for spinning up the web app for local development)
  - `src/` - contains all the sample application code which you can modify or swap out for your own application
  - `tests/` - test for the application, written using [Playwright](https://playwright.dev/)

### Infrastructure as Code (provisioning and deploying infrastructure on Azure, programmatically)

>   - `infra/` - contains all infrastructure as code (IaC) assets written in [Bicep](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/overview?tabs=bicep); includes logic to set up all components we need to set up the application on Azure, wire everything up securely, and monitor application health, performance and usage!

Reviewer: callout on support for Terraform?

### CI/CD
  - `.github/` - contains a GitHub Actions workflow to set up a CI/CD pipeline that runs on every new commit to the repo

## **End-to-end support:** Move code to cloud in a single step!
### If you want to follow along in this section and don't already have the Azure Developer CLI installed, check out [these instructions](https://aka.ms/azure-dev/install-instructions?source=serverless-september) to install it on Windows, Linux or MacOS in your favorite terminal!

So now that we've gone over what this template contains on GitHub, let's pull this template code down to our local machine, set it up for local development, provision the right infrastructure, and deploy the code on Azure in a **single step**.

> When designing the CLI, we wanted the experience to be both flexible and non-magical (no side effects, easy to understand). That being said, we're going to run this all with `azd up` but you could alternatively run a series of three commands and the outcome would be the same - `azd init --template Azure-Samples/todo-python-mongo-swa-func` (to pull the code down to your machine), `azd provision` (to provision infrastructure) and then `azd deploy` (to deploy application code on Azure). Choose your own adventure!

Reviewer 2: By using "that being said, " it feels like you are saying there is magic.  You might want to replace it with "Next" or something like that.

![](img/single-step.png)

> So let's walk through it. On running `azd up --template Azure-Samples/todo-python-mongo-swa-func`, I'm prompted for a couple pieces of information as part of the `azd init` process being run under the hood:

Reviewer 2: ```suggestion
So let's walk through it. On running `azd up -t todo-python-mongo-swa-func`, I'm prompted for a couple pieces of information as part of the `azd init` process being run under the hood:
```

    - **An environment name** - the prefix for the resource group that will be created to hold all Azure resources
    - **An Azure region** - the Azure location where your resources will be deployed

>     - **An Azure subscription** - the Azure location where your resources will be deployed

Reviewer: the Azure account/subscription used for deploying resources

Author: Yeah - I copied this from the docs for consistency but I agree. It seems like a mistake that the description for region and sub are the same.

@puicchan Thoughts here?

Reviewer 3: In readme, it's 
![image](https://user-images.githubusercontent.com/8701892/191550568-3a4e32a7-ea2d-41b1-a9db-cd425dc9cf4d.png)

Author: Ah - my bad. I think we should still consider amending that text. We can chat.

![](img/start-up.png)

> Once that information is provided, `azd` will pull down the code from GitHub and create a `.azure/` in the project root that contains all Azure Developer CLI environment information that you just entered. This directory will be important when it comes time to provision and deploy infrastructure in the next step in the `up` process.

Reviewer 2: ```suggestion
Once that information is provided, `azd` will pull down the code from GitHub and create a `.azure/` directory in the project root that contains all Azure Developer CLI environment information that you just entered. This directory will be important when it comes time to provision and deploy infrastructure in the next step in the `up` process.
```

The next step here is provisioning. `azd` is running `azd provision` on your behalf and leveraging the IaC assets in the `.infra/` directory in the project. As the tool works to provision, you'll see an output of each resource (name alongside a unique identifier which you can use to reference back to the Azure Portal, if you want)
![](img/provision-up.png)

Finally, the final step here in running `azd up` is deployment. `azd` is running `azd deploy` and deploying the application code to the resources that we're provisioned in the previous phase of the process. Once this has completed, you'll be able to click on two different endpoint URLs - one for the back-end and one for the front-end. 
![](img/deploy-up.png)

> The back-end endpoint (`service api`) hosts the specification for the API via the `openapi.json` file that's also in the root of the project template. You can explore the endpoints that are available in the web app here. 

Reviewer 2: We don't hyphen frontend and backend in our docs.

> The back-end endpoint (`service api`) hosts the specification for the API via the `openapi.json` file that's also in the root of the project template. You can explore the endpoints that are available in the web app here. 

Reviewer 2: ```suggestion
The back-end endpoint (`service api`) hosts the specification for the API via the `openapi.yaml` file that's also in the root of the project template. You can explore the endpoints that are available in the web app here. 
```

![](img/backend.png)

The front-end endpoint (`service web`) hosts a fully-fledged and functional ToDo web app with a UI, Cosmos DB for the database and Key Vault for application secrets. This isn't just application hosting. It's really everything you need to be successful and productive, all set up on your behalf by the Azure Developer CLI.
![](img/frontend.png)

...and that's it! We've successfully deployed our application on Azure! 

But there's more!

## Best practices: Monitoring and CI/CD!

> In my opinion, it's not enough to _just_ set up the application on Azure! I want to know that my web app is performant and serving my users reliably! I also want to make sure that I'm not inadvertently breaking my application as I continue to make changes to it. Thankfully, the Developer CLI also handles all of this via two additional commands - `azd monitor` and `azd pipeline config`.

Reviewer 2: ```suggestion
In my opinion, it's not enough to _just_ set up the application on Azure! I want to know that my web app is performant and serving my users reliably! I also want to make sure that I'm not inadvertently breaking my application as I continue to make changes to it. Thankfully, the Azure Developer CLI also handles all of this via two additional commands - `azd monitor` and `azd pipeline config`.
```

Reviewer 2: Prob want to add Azure to anywhere you have Developer CLI
